# CAR PLATES DETECTION 
Names and ID's:

1- Omar Khaled - ID:94105

2- Mostafa Tarek - ID:94071

3- Mohamed Ezz - ID:94303

4- Mohamed Omar - ID:94114

5- Basel Ahmed - ID:98052

#Install Dependencies

We will Install YOLO v7 Dependencies to begin usage of our model.

In [ ]:
import torch
torch.cuda.empty_cache()

In [ ]:
torch.cuda.memory_summary(device=None, abbreviated=False)

'|===========================================================================|\n|                  PyTorch CUDA memory summary, device ID 0                 |\n|---------------------------------------------------------------------------|\n|            CUDA OOMs: 0            |        cudaMalloc retries: 0         |\n|===========================================================================|\n|        Metric         | Cur Usage  | Peak Usage | Tot Alloc  | Tot Freed  |\n|---------------------------------------------------------------------------|\n| Allocated memory      |       0 B  |       0 B  |       0 B  |       0 B  |\n|       from large pool |       0 B  |       0 B  |       0 B  |       0 B  |\n|       from small pool |       0 B  |       0 B  |       0 B  |       0 B  |\n|---------------------------------------------------------------------------|\n| Active memory         |       0 B  |       0 B  |       0 B  |       0 B  |\n|       from large pool |       0 B  |       0 B  |

In [ ]:
!nvidia-smi

Mon Jan  2 08:19:07 2023       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 460.32.03    Driver Version: 460.32.03    CUDA Version: 11.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla T4            Off  | 00000000:00:04.0 Off |                    0 |
| N/A   69C    P0    30W /  70W |      3MiB / 15109MiB |      4%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-------

In [ ]:
!git clone https://github.com/SkalskiP/yolov7.git
%cd yolov7
!git checkout fix/problems_associated_with_the_latest_versions_of_pytorch_and_numpy
!pip install -r requirements.txt

Cloning into 'yolov7'...
remote: Enumerating objects: 579, done.
remote: Counting objects: 100% (7/7), done.
remote: Compressing objects: 100% (6/6), done.
remote: Total 579 (delta 1), reused 3 (delta 1), pack-reused 572
Receiving objects: 100% (579/579), 38.18 MiB | 23.34 MiB/s, done.
Resolving deltas: 100% (283/283), done.
/content/yolov7
Branch 'fix/problems_associated_with_the_latest_versions_of_pytorch_and_numpy' set up to track remote branch 'fix/problems_associated_with_the_latest_versions_of_pytorch_and_numpy' from 'origin'.
Switched to a new branch 'fix/problems_associated_with_the_latest_versions_of_pytorch_and_numpy'
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 1.6 MB 6.9 MB/s 


# Download Correctly Formatted Custom Data

Next, we'll download our dataset in the right format. Use the `YOLOv7 PyTorch` export. Note that this model requires YOLO TXT annotations, a custom YAML file, and organized directories. 

Here we Used Roboflow.

In [ ]:
!pip install roboflow
from roboflow import Roboflow
rf = Roboflow(api_key="ZK8jie9hxh6VvLZAv5Gw")
project = rf.workspace("letters").project("letete")
dataset = project.version(1).download("yolov7")

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 45 kB 2.7 MB/s 
     |████████████████████████████████| 178 kB 13.8 MB/s 
     |████████████████████████████████| 138 kB 64.7 MB/s 
     |████████████████████████████████| 67 kB 7.2 MB/s 
     |████████████████████████████████| 54 kB 3.3 MB/s 
     |████████████████████████████████| 62 kB 1.8 MB/s 
  Created wheel for wget: filename=wget-3.2-py3-none-any.whl size=9674 sha256=b960d590c4c0042f108fdc10a35d95b9dd0ddf37291b17111336e38b91724c7e
  Stored in directory: /root/.cache/pip/wheels/bd/a8/c3/3cf2c14a1837a4e04bd98631724e81f33f462d86a1d895fae0
Successfully built wget
  Attempting uninstall: urllib3
    Found existing installation: urllib3 1.24.3
    Uninstalling urllib3-1.24.3:
      Successfully uninstalled urllib3-1.24.3
  Attempting uninstall: requests
    Found existing installation: requests 2.23.0
    Uninstalling requests-2.23.0:
      Succes

loading Roboflow workspace...
loading Roboflow project...


Extracting Dataset Version Zip to letete-1 in yolov7pytorch:: 100%|██████████| 1134/1134 [00:00<00:00, 1479.49it/s]


# Begin Custom Training

We're ready to start custom training.

You can add functions or Edit functions depending on your Model and data set.
Here we Used 230 Epochcs with Batch size of 15.

In [ ]:
%cd /content/yolov7
!wget https://github.com/WongKinYiu/yolov7/releases/download/v0.1/yolov7_training.pt

/content/yolov7
--2023-01-02 08:19:42--  https://github.com/WongKinYiu/yolov7/releases/download/v0.1/yolov7_training.pt
Resolving github.com (github.com)... 192.30.255.113
Connecting to github.com (github.com)|192.30.255.113|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/511187726/13e046d1-f7f0-43ab-910b-480613181b1f?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20230102%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20230102T081830Z&X-Amz-Expires=300&X-Amz-Signature=7bd5ad307c34b33189b16b64783f6246e331120226ae2aee10b497a7a6f9929c&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=511187726&response-content-disposition=attachment%3B%20filename%3Dyolov7_training.pt&response-content-type=application%2Foctet-stream [following]
--2023-01-02 08:19:42--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/511187726/13e046d1-f7f0-43ab-9

In [ ]:
# run this cell to begin training
%cd /content/yolov7
!python train.py --batch-size 15 --epochs 150 --data {dataset.location}/data.yaml --weights 'yolov7_training.pt' --device 0 


/content/yolov7
YOLOR 🚀 b2a7de9 torch 1.13.0+cu116 CUDA:0 (Tesla T4, 15109.75MB)

Namespace(adam=False, artifact_alias='latest', batch_size=15, bbox_interval=-1, bucket='', cache_images=False, cfg='', data='/content/yolov7/letete-1/data.yaml', device='0', entity=None, epochs=150, evolve=False, exist_ok=False, freeze=[0], global_rank=-1, hyp='data/hyp.scratch.p5.yaml', image_weights=False, img_size=[640, 640], label_smoothing=0.0, linear_lr=False, local_rank=-1, multi_scale=False, name='exp', noautoanchor=False, nosave=False, notest=False, project='runs/train', quad=False, rect=False, resume=False, save_dir='runs/train/exp3', save_period=-1, single_cls=False, sync_bn=False, total_batch_size=15, upload_dataset=False, v5_metric=False, weights='yolov7_training.pt', workers=8, world_size=1)
tensorboard: Start with 'tensorboard --logdir runs/train', view at http://localhost:6006/
hyperparameters: lr0=0.01, lrf=0.1, momentum=0.937, weight_decay=0.0005, warmup_epochs=3.0, warmup_momentum=0.8, 

# Evaluation

We can evaluate the performance of our custom training using the provided evalution script and run inference on images

In [ ]:
!python detect.py --weights runs/train/exp/weights/best.pt --conf 0.2 --source {dataset.location}/test/images

Namespace(agnostic_nms=False, augment=False, classes=None, conf_thres=0.2, device='', exist_ok=False, img_size=640, iou_thres=0.45, name='exp', no_trace=False, nosave=False, project='runs/detect', save_conf=False, save_txt=False, source='/content/yolov7/letete-1/test/images', update=False, view_img=False, weights=['runs/train/exp/weights/best.pt'])
YOLOR 🚀 b2a7de9 torch 1.13.0+cu116 CUDA:0 (Tesla T4, 15109.75MB)

Traceback (most recent call last):
  File "/content/yolov7/utils/google_utils.py", line 26, in attempt_download
    assets = [x['name'] for x in response['assets']]  # release assets
KeyError: 'assets'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "detect.py", line 196, in <module>
    detect()
  File "detect.py", line 34, in detect
    model = attempt_load(weights, map_location=device)  # load FP32 model
  File "/content/yolov7/models/experimental.py", line 251, in attempt_load
    attempt_download(w)
  File "/

In [ ]:
import glob
from IPython.display import Image, display
i = 0
limit = 10000 
for imageName in glob.glob('/content/yolov7/runs/detect/exp/*.jpg'): 
    if i < limit:
      display(Image(filename=imageName))
      print("\n")
    i = i + 1

# Reparameterize for Inference

https://github.com/WongKinYiu/yolov7/blob/main/tools/reparameterization.ipynb

#  Deployment

To deploy, you'll need to export your weights and save them to use later.

In [ ]:
!zip -r export.zip runs/detect
!zip -r export.zip runs/train/exp/weights/best.pt
!zip export.zip runs/train/exp/*

updating: runs/detect/ (stored 0%)
updating: runs/detect/exp/ (stored 0%)
updating: runs/detect/exp2/ (stored 0%)
  adding: runs/detect/exp4/ (stored 0%)
  adding: runs/detect/exp3/ (stored 0%)
	zip warning: name not matched: runs/train/exp/weights/best.pt

zip error: Nothing to do! (try: zip -r export.zip . -i runs/train/exp/weights/best.pt)
updating: runs/train/exp/events.out.tfevents.1672433926.f3f93ec34a6e.688.0 (deflated 48%)
updating: runs/train/exp/hyp.yaml (deflated 44%)
updating: runs/train/exp/opt.yaml (deflated 46%)
updating: runs/train/exp/train_batch0.jpg (deflated 1%)
updating: runs/train/exp/train_batch1.jpg (deflated 2%)
updating: runs/train/exp/train_batch2.jpg (deflated 3%)
updating: runs/train/exp/train_batch3.jpg (deflated 3%)
updating: runs/train/exp/train_batch4.jpg (deflated 3%)
updating: runs/train/exp/train_batch5.jpg (deflated 2%)
updating: runs/train/exp/train_batch6.jpg (deflated 3%)
updating: runs/train/exp/train_batch7.jpg (deflated 3%)
updating: runs/trai

# THAT'S IT

We trained our Object Detection Model Using YOLOV7 and it achieved 0.94 mAP! 

We can use the model to read the plate number with high performance.